File: lib/aws/ec2/securityGroup.ts
// Copyright 2016 Pulumi, Inc. All rights reserved.

import {VPC} from './vpc';
import {VPCPeeringConnection} from './vpcPeeringConnection';
import * as cloudformation from '../cloudformation';

// An Amazon EC2 security group.
// @name: aws/ec2/securityGroup
// @website: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-security-group.html
export class SecurityGroup
        extends cloudformation.Resource
        implements SecurityGroupProperties {

    public readonly groupDescription: string;
    public readonly vpc?: VPC;
    public securityGroupEgress?: SecurityGroupRule[];
    public securityGroupIngress?: SecurityGroupRule[];

    constructor(name: string, args: SecurityGroupProperties) {
        super({
            name: name,
            resource:  "AWS::EC2::SecurityGroup",
        });
        this.groupDescription = args.groupDescription;
        this.vpc = args.vpc;
        this.securityGroupEgress = args.securityGroupEgress;
        this.securityGroupIngress = args.securityGroupIngress;
    }
}

export interface SecurityGroupProperties extends cloudformation.TagArgs {
    // Description of the security group.
    readonly groupDescription: string;
    // The VPC in which this security group resides.
    readonly vpc?: VPC;
    // A list of Amazon EC2 security group egress rules.
    securityGroupEgress?: SecurityGroupRule[];
    // A list of Amazon EC2 security group ingress rules.
    securityGroupIngress?: SecurityGroupRule[];
}

// An EC2 Security Group Rule is an embedded property of the SecurityGroup.
export interface SecurityGroupRule {
    // An IP protocol name or number.
    ipProtocol: string;
    // Specifies a CIDR range.
    cidrIp?: string;
    // The start of port range for the TCP and UDP protocols, or an ICMP type number. An ICMP type number of `-1`
    // indicates a wildcard (i.e., any ICMP type number).
    fromPort?: number;
    // The end of port range for the TCP and UDP protocols, or an ICMP code. An ICMP code of `-1` indicates a wildcard
    // (i.e., any ICMP code).
    toPort?: number;
}

File: lib/aws/idl/ec2/securityGroup.go
// Copyright 2016-2017, Pulumi Corporation.  All rights reserved.

package ec2

import (
	"github.com/pulumi/lumi/pkg/resource/idl"
)

// A SecurityGroup is an Amazon EC2 Security Group.  For more information, see
// http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-security-group.html.
type SecurityGroup struct {
	idl.NamedResource
	// A required description about the security group.
	GroupDescription string `lumi:"groupDescription,replaces"`
	// An optional name for the security group.  If you don't specify one, a unique physical ID will be generated and
	// used instead.  If you specify a name, you cannot perform updates that require replacement of this resource.  You
	// can perform updates that require no or some interruption.  If you must replace the resource, specify a new name.
	GroupName *string `lumi:"groupName,optional,replaces"`
	// The VPC in which this security group resides (or blank if the default VPC).
	VPC *VPC `lumi:"vpc,optional,replaces"`
	// A list of Amazon EC2 security group egress rules.
	SecurityGroupEgress *[]SecurityGroupRule `lumi:"securityGroupEgress,optional"`
	// A list of Amazon EC2 security group ingress rules.
	SecurityGroupIngress *[]SecurityGroupRule `lumi:"securityGroupIngress,optional"`
	// The group ID of the specified security group, such as `sg-94b3a1f6`.
	GroupID string `lumi:"groupID,out"`
}

// A SecurityGroupRule describes an EC2 security group rule embedded within a SecurityGroup.
type SecurityGroupRule struct {
	// The IP name or number.
	IPProtocol string `lumi:"ipProtocol"`
	// Specifies a CIDR range.
	CIDRIP *string `lumi:"cidrIp,optional"`
	// The start of port range for the TCP and UDP protocols, or an ICMP type number.  An ICMP type number of `-1`
	// indicates a wildcard (i.e., any ICMP type number).
	FromPort *float64 `lumi:"fromPort,optional"`
	// The end of port range for the TCP and UDP protocols, or an ICMP code.  An ICMP code of `-1` indicates a
	// wildcard (i.e., any ICMP code).
	ToPort *float64 `lumi:"toPort,optional"`
}
